### ISSUE TITLE
FRAGMOS - Acc DRR Project - PortfolioPriceReturnTerms

### ISSUE THREAD

**Issue by JBZ-Fragmos** [finos/common-domain-model#2745]
**BUSINESS NEED**

To represent a price return swap on Basket underlier, with **specific features** below :
- booked in the Risk System as a “portfolio” of **multiple individual price return legs on Single underlier**
- **frequent** position changes : increase / decrease / substitutions 
- **net position reversal** may happen (e.g. when decrease size exceeds prior position, then initial long position would turn into net short position)
- may need to have s**pecific financing terms per underlier leg** (that is optional i.e. some cases only need one such leg, as in “classic” swap)

Eventually, **operational issues** arise because the product needs to behave as either a « Basket» or as « multiple Individual legs» in parallel, depending on the business need e.g. FO hedge, DRR, MO cash calculation, MO confirmations – that without internal disprepancy between all these complementary usages.

**MODELLING PROPOSAL**

**related PR :** https://github.com/finos/common-domain-model/pull/2744

- As regards the product per se : create new type portfolioPriceReturn at priceReturn node, that is where the right balance is for an aggregated payout to hold an aggregated position at Basket level, while limiting redundancies and also permitting to associate an interestRate leg per return leg, if need be
- Additional refactoring are proposed, which would benefit to any return swap that mainly consist in :
  - **moving** existing objects used to represent **valuation prices** i.e. refactoring where valuation prices are placed
  - also **rename** them e.g. from "valuationDates" to "valuationTerms" (because related type actually contains more terms than only "dates", hence related name should not only focus on "date" cf. see "OTHER REFACTORING section below)
  - add referencing properties to these items (metadata reference) so that valuation price could reference Price from TradeLot->PriceQuantity
- As regards requirements related to position lifecycle – will be addressed in a separate Issue / separate PR

**CHANGES DESCRIPTION**

- **_MAIN_**

![image](https://github.com/finos/common-domain-model/assets/92978493/729da3c1-ee04-4383-bba9-eb01529cf519)

![image](https://github.com/finos/common-domain-model/assets/92978493/d4f508dd-ff9b-4b3c-a470-4e1756c580ac)

![image](https://github.com/finos/common-domain-model/assets/92978493/ce33d580-822e-4169-962b-9e433c2f26ba)

![image](https://github.com/finos/common-domain-model/assets/92978493/54848298-e361-4273-9c96-970efe069b77)

- _**OTHER REFACTORING**_

in addition, for consistency purposes in regards of attribute names proposed as part of such as "initialValuationPrice", etc. 
![image](https://github.com/finos/common-domain-model/assets/92978493/b303c25b-2b74-4ba3-9ead-0a3b2ef6b0e6)

for the avoidance of doubts, the terms removed below are not removed from the model, they are moved at ReturnTerms level, 
![image](https://github.com/finos/common-domain-model/assets/92978493/4a805ecc-b809-4aac-a50c-70b6d6ad7d9f)

![image](https://github.com/finos/common-domain-model/assets/92978493/23bec378-824d-45c6-9ee6-3279565de66f)

**Comment by lolabeis** [comment]
@JBZ-Fragmos Before I comment further on the Pull Request, can you please clarify on the design:

It looks like you're building multiple cardinality redundancy:
- On the one hand your `portfolioPriceReturnTerms` attribute is set as `(0..*)`
- On the other you're allowing a single `underlier` to handle multiple `basketConstituent`

If the latter is the chosen representation, I would have expected `portfolioPriceReturnTerms` to be `(0..1)`.

**Comment by JBZ-Fragmos** [comment]
@lolabeis 
as documented in related GitHub Issue and also during works in cessions, this is on purpose, i mean this is exactly what we need here in terms of business requirements, hence need to create new types, etc.
effectively, we want in same product to have BOTH 

- **multiple** single return per each single underlier 
- AND one Basket at "net" level, which **multiple** components are these single underliers

Lionel has been involved as you had requested, so you may want to liaise directly with him also to help answering your questions

**Comment by lolabeis** [comment]
Ok, I think I get the redundancy design. So overall basket `underlier` in `PerformancePayout`, and individual `underliers` in a multiplicity of `portfolioPriceReturnTerms` objects, both pointing to the same data in the external `PriceQuantity` object.

Couple of follow-up comments:
* It looks like you're aligning `PortfolioPriceReturnTerms` onto `ReturnTerms` (+/- some attributes), which is good. Consider renaming the former as `PortfolioReturnTerms`, since it contains both price and dividend return terms.
* By the same argument, consider repositioning `portfolioReturnTerms` **alongside** `returnTerms` in `PerformancePayout` (still with multiple cardinality) rather than **in it** - possibly with a choice condition between `returnTerms` and `portfolioReturnTerms`?
* Why position `financingTerms` inside `PortfolioPriceReturnTerms` instead of simply using `interestRatePayout` in `Payout`? As I understand it, you will need to create multiple `portfolioPriceReturnTerms` attributes, each with its own `financingTerms`. So in terms of overall `Payout` object size it would be the same if you were creating multiple `interestRatePayout` attributes. If you did that, you would still be able to link the quantity of each `interestRatePayout` to the quantity of the corresponding `PortfolioPriceReturnTerms` object, in the same way as is currently done for single-name equity swaps
* You are proposing some further refactoring, which has merit. However you will be constrained by having to keep CDM Prod (which is what is used in DRR) backward-compatible, so removing or renaming attributes is a no-go. You may use the `[deprecated]` annotation to keep those attributes in place while introducing new ones, but I would advise against doing any  non-critical refactoring and for saving those for CDM Dev. It also means that your Prod vs Dev Pull Requests will be shaped differently, rather than a simple copy/paste.

**Comment by Oblongs** [comment]
Here is the deck that was presented at the DPBE on 27 February:
[Price Return Swap v3 DPBE.pptx](https://github.com/finos/common-domain-model/files/14511876/Price.Return.Swap.v3.DPBE.pptx)

**Comment by dshoneisda** [comment]
I am hoping we can move this forward in Monday's meeting. To the question above we should where possible try to do enough to support DRR requirements of JPM/Pictet, and note further development/design improvement for a later date if necessary

**Comment by JBZ-Fragmos** [comment]
TAKING INTO ACCOUNT ALL ABOVE DISCUSSION AND FEEDBACK 

**HEREUNDER refactored proposal**

![image](https://github.com/finos/common-domain-model/assets/92978493/c70cc69d-5f06-4d06-862f-f0facd140782)

![image](https://github.com/finos/common-domain-model/assets/92978493/2abb1060-5b8e-44a1-8a32-4d34bf241124)

![image](https://github.com/finos/common-domain-model/assets/92978493/5edcc713-9e11-4b64-b750-c763b3a95604)

![image](https://github.com/finos/common-domain-model/assets/92978493/dbfd64a0-585e-4aa7-83e9-6757ba490b23)

**Comment by tomhealey-icma** [comment]
My initial comment is that BasketConstituent is misplaced. As defined it includes economic characteristics that belong at the economicTerms level for attributes or provisions that effect the overall basket and at the payout level for those items that effect payout characteristics. 

The description of the problem appears to describe a portfolio (not considered a tradable product) so if these terms are not specific the definition of a product then they belong in the portfolio. Baskets are considered tradable products. 

I'll continue to review but please clarify the mixed use portfolio and basket terms.

**Comment by JBZ-Fragmos** [comment]
@tomhealey-icma

the fact BasketConstituent is of type Product (which effectively mean that it may include economicTerms etc. as you say above) is **already** in the model today

if you would like to discuss it, that should be part of another Issue, mostly as part of other good discussions we've had about "Asset", etc.

for now, we shall focus on changes related to "portfolioReturnSwap" 
and having Product as a BasketConsituent is not a change from "portfolioReturnSwap" 

indeed Baset refactoring proposed as part of "portfolioReturnSwap" proposal, **only "extends (_existing_) Product"** with other attributes required to better represent net aggregation Price and Quantity at Basket level, but again, that is not touching current model where Product is BasketConsitutent

![image](https://github.com/finos/common-domain-model/assets/92978493/12c60edb-5ccd-40fa-977d-c05c6ec0818c)

**Comment by JBZ-Fragmos** [comment]
As an indication, have just pushed an update with 2 changes : 

**condition is slitghly modified as below :** 
![image](https://github.com/finos/common-domain-model/assets/92978493/1fd5d71e-bc33-485a-a3e5-acfd251c74c0)

**payerReceiver is added to PortfolioReturnTerms** (because depending on strategy and business case, the payerReceiver of the whole structure, that is defined at PerformancePayout level, may not necessarily be the same as inside individual legs, moreover for some strategy, it will necessarily vary inside the individual legs)
![image](https://github.com/finos/common-domain-model/assets/92978493/546dc220-5481-46dd-8d4a-368aafae3e41)
